shuup/core/models/_orders.py

```diff
@@ -1106,5 +1163,47 @@ def get_available_payment_methods(self):
             if m.is_available_for(self)
         ]
 
+    # check whether user can be able to change status of order
+    def validate_order_status_change(self, next_status: OrderStatus):
+        valid_statuses = [OrderStatusRole.NONE, OrderStatusRole.INITIAL, OrderStatusRole.PROCESSING,
+                          OrderStatusRole.COMPLETE, OrderStatusRole.CANCELED]
+
+        if not next_status or next_status not in valid_statuses \
+                or next_status == OrderStatusRole.INITIAL:
+            return False
+
+        curr_role = self.status.role
+        # valid follow of change status:
+        # 1. `INITIAL -> PROCESSING -> COMPLETE`
+        # 2. `INITIAL -> PROCESSING -> CANCELED`
+        # 3. `INITIAL -> CANCELED`
+        if curr_role in [OrderStatusRole.COMPLETE, OrderStatusRole.CANCELED] \
+                or (curr_role == OrderStatusRole.PROCESSING and next_status == OrderStatusRole.INITIAL):
+            return False
+
+        return True
+
+    def change_status(self, next_status: OrderStatus, user: User = None, description: str = None):
+        # validate next_status is valid or not
+
+        if not self.validate_order_status_change(next_status):
```

Comment:
this method should check whether `next_status` is inside `self.status.allowed_next_statuses.all()`, just like that